openfl-gramine/MANUAL.md

```diff
@@ -0,0 +1,189 @@
+- During plan initialization we need data to initialize the model. so at least one collaborator should be in data.yaml and its data should be available. cols.yaml may be empty at first
+During cert sign request generation cols.yaml on collaborators remain empty, data.yaml is extended if needed. On aggregator, cols.yaml are updated during signing procedure, data.yaml remains unmodified
+- `error: Disallowing access to file '/usr/local/lib/python3.8/__pycache__/signal.cpython-38.pyc.3423950304'; file is not protected, trusted or allowed.`
+## Prerquisites
+Building machine:
+- OpenFL
+- Docker should be installed, user included in Docker group
+
+Machines that will run an Aggregator and Collaborator containers should have the following:
+- SGX enebled in BIOS
+- Ubuntu with Linux kernel 5.11+
+- ? SGX drivers, it is built in kernel: `/dev/sgx_enclave`
+- [aesmd service](https://github.com/intel/linux-sgx) (`/var/run/aesmd/aesm.socket`)
+This is a short list, see more in Gramine docs.
+
+## Workflow
+The user will mainly interact with OpenFL CLI, docker CLI, and other command-line tools. But the user is also expected to modify plan.yaml file and Python code under workspace/src folder to set up an FL Experiment.
+### On the building machine (Data Scientist's node):
+1. As usual, **create a workspace**: 
+```
+export WORKSPACE_NAME=my_sgx_federation_workspace
+export TEMPLATE_NAME=torch_cnn_histology
+
+fx workspace create --prefix $WORKSPACE_NAME --template $TEMPLATE_NAME
+cd $WORKSPACE_NAME
+```
+Modify the code and the plan.yaml, set up your training procedure. </br>
+Pay attention to the following: 
+- make sure the data loading code reads data from ./data folder inside the workspace
+- if you download data (development scenario) make sure your code first checks if data exists, as connecting to the internet from an enclave may be problematic.
+- make sure you do not use any CUDA driver-dependent packages
+
+Default workspaces (templates) in OpenFL differ in their data downloading procedures. Workspaces with data loading flow that do not require changes to run with Gramine include:
+- torch_unet_kvasir
+- torch_cnn_histology
+- keras_nlp
+
+2. **Initialize the experiment plan** </br> 
+Find out the FQDN of the aggregator machine and use it for plan initialization.
+For example, on Unix-like OS try the following command:
+```
+hostname --all-fqdns | awk '{print $1}'
+```
+Then pass the result as `AGG_FQDN` parameter to:
+```
+fx plan initialize -a $AGG_FQDN
+```
+
+3. (Optional) **Generate a signing key** on the building machine if you do not have one.</br>
+It will be used to calculate hashes of trusted files. If you plan to test the application without SGX (gramine-direct) you also do not need a signer key.
+```
+export KEY_LOCATION=.
+
+openssl genrsa -3 -out $KEY_LOCATION/key.pem 3072
```

Comment:
This week the gramine project added a function to generate the signing key with the cryptography package: https://github.com/gramineproject/gramine/commit/3def085a9689d6642d2ad9d86e962e227d8b0ca4

We should reuse/copy this functionality to avoid adding OpenSSL as another dependency

Reply:
I've just checked and this command is not in the package available via apt yet